src/parser/on_demand/parse_to_structure.ts

```diff
@@ -0,0 +1,143 @@
+import { type Code } from '../../code';
+import { type BSONElement, getSize, parseToElements } from './parse_to_elements';
+
+/** @internal */
+const DEFAULT_REVIVER: BSONReviver = (
+  _bytes: Uint8Array,
+  _container: Container,
+  _element: BSONElement
+) => null;
+
+/** @internal */
+function parseToElementsToArray(bytes: Uint8Array, offset?: number | null): BSONElement[] {
+  const res = parseToElements(bytes, offset);
+  return Array.isArray(res) ? res : [...res];
+}
+
+/** @internal */
+type ParseContext = {
+  elementOffset: number;
+  elements: BSONElement[];
+  container: Container;
+  previous: ParseContext | null;
+};
+
+/**
+ * @experimental
+ * @public
+ * A union of the possible containers for BSON elements.
+ *
+ * Depending on kind, a reviver can accurately assign a value to a name on the container.
+ */
+export type Container =
+  | {
+      dest: Record<string, unknown>;
+      kind: 'object';
+    }
+  | {
+      dest: Map<string, unknown>;
+      kind: 'map';
+    }
+  | {
+      dest: Array<unknown>;
+      kind: 'array';
+    }
+  | {
+      dest: Code;
+      kind: 'code';
+    }
+  | {
+      kind: 'custom';
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      dest: any;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      [key: string]: any;
+    };
+
+/**
+ * @experimental
+ * @public
+ */
+export type BSONReviver = (
+  bytes: Uint8Array,
+  container: Container,
+  element: BSONElement
+) => Container | null;
+
+/**
+ * @experimental
+ * @public
+ */
+export function parseToStructure<
+  TRoot extends Container = {
+    dest: Record<string, unknown>;
+    kind: 'object';
+  }
+>(
+  bytes: Uint8Array,
+  startOffset?: number | null,
+  pRoot?: TRoot | null,
+  pReviver?: BSONReviver | null
```

Comment:
(See other comment, comes from Java)

I can't write `root ??= ...` because TS warns that root may have been instantiated with a different type. Maybe I'm doing something wrong but I think the problem is TS can't prevent this invocation:

```ts
parseToStructure<{ kind: 'custom'; dest: number }>(bytes);
```
Making the thing returned different from the type given. But this is not the intended use, it is supposed to automatically determine the type:
```ts
parseToStructure(bytes, undefined, { kind: 'custom', dest: 0 });
```
Maybe I have to use overrides 🤔

Reply:
Can you not just use default values? 

```typescript
export function parseToStructure<
  TRoot extends Container = {
    dest: Record<string, unknown>;
    kind: 'object';
  }
>(
  bytes: Uint8Array,
  startOffset?: number | null,
  root: TRoot | null = {
    kind: 'object',
    dest: Object.create(null) as Record<string, unknown>
  },
  reviver: BSONReviver | null = DEFAULT_REVIVER
)
```

Reply:
No default values only work for undefined, I wanted to support null to match our `== null` style code for options defaulting. Also, that doesn't fix the type issue: `'TRoot' could be instantiated with a different subtype of constraint 'Container'`. As a TS library author I want to use TRoot to automagically determine the return type based on the type of root so that users never use `parseToStructure<XXX>` type parameterization syntax. I just think there isn't a nice way to allow one without the other, so new variable it is. Open to ideas if there is a way